pkg/compactor/compactor.go

```diff
@@ -0,0 +1,291 @@
+	if err != nil {
+		cancelCtx()
+		return nil, errors.Wrap(err, "failed to create TSDB compactor")
+	}
+
+	return newCompactor(ctx, cancelCtx, compactorCfg, storageCfg, bucketClient, tsdbCompactor, logger, registerer)
+}
+
+func newCompactor(
+	ctx context.Context,
+	cancelCtx context.CancelFunc,
+	compactorCfg Config,
+	storageCfg cortex_tsdb.Config,
+	bucketClient objstore.Bucket,
+	tsdbCompactor tsdb.Compactor,
+	logger log.Logger,
+	registerer prometheus.Registerer,
+) (*Compactor, error) {
+	c := &Compactor{
+		compactorCfg:  compactorCfg,
+		storageCfg:    storageCfg,
+		logger:        logger,
+		bucketClient:  bucketClient,
+		tsdbCompactor: tsdbCompactor,
+		ctx:           ctx,
+		cancelCtx:     cancelCtx,
+		compactionRunsStarted: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "cortex_compactor_runs_started_total",
+			Help: "Total number of compaction runs started.",
+		}),
+		compactionRunsCompleted: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "cortex_compactor_runs_completed_total",
+			Help: "Total number of compaction runs successfully completed.",
+		}),
+		compactionRunsFailed: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "cortex_compactor_runs_failed_total",
+			Help: "Total number of compaction runs failed.",
+		}),
+	}
+
+	// Register metrics.
+	if registerer != nil {
+		registerer.MustRegister(c.compactionRunsStarted, c.compactionRunsCompleted, c.compactionRunsFailed)
+	}
+
+	// Start the compactor loop.
+	c.runner.Add(1)
+	go c.run()
+
+	return c, nil
+}
+
+// Shutdown the compactor and waits until done. This may take some time
+// if there's a on-going compaction.
+func (c *Compactor) Shutdown() {
+	c.cancelCtx()
+	c.runner.Wait()
+}
+
+func (c *Compactor) run() {
+	defer c.runner.Done()
+
+	// Run an initial compaction before starting the interval.
+	c.compactUsersWithRetries(c.ctx)
+
+	ticker := time.NewTicker(c.compactorCfg.CompactionInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			c.compactUsersWithRetries(c.ctx)
+		case <-c.ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *Compactor) compactUsersWithRetries(ctx context.Context) {
+	retries := util.NewBackoff(ctx, util.BackoffConfig{
+		MinBackoff: c.compactorCfg.retryMinBackoff,
+		MaxBackoff: c.compactorCfg.retryMaxBackoff,
+		MaxRetries: c.compactorCfg.CompactionRetries,
+	})
+
+	c.compactionRunsStarted.Inc()
+
+	for retries.Ongoing() {
+		if err := c.compactUsers(ctx); err == nil {
+			c.compactionRunsCompleted.Inc()
+			return
+		}
+
+		retries.Wait()
+	}
+
+	c.compactionRunsFailed.Inc()
+}
+
+func (c *Compactor) compactUsers(ctx context.Context) error {
+	level.Info(c.logger).Log("msg", "discovering users from bucket")
+	users, err := c.discoverUsers(ctx)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to discover users from bucket", "err", err)
+		return err
```

Comment:
Since error is not used by the caller, can we return bool instead? It's just weird to see error being logged, but also returned.